test/show.jl

```diff
@@ -755,6 +755,69 @@ end
 
 @test startswith(sprint(show, typeof(x->x), context = :module=>@__MODULE__), "var\"")
 
+# PR 53719
+module M53719
+    f = x -> x + 1
+    function foo(x)
```

Comment:
This should probably also add some tests for functions with `#` in their names, which is not that much of a corner case thanks to macros:
```julia
julia> macro foo()
           return quote
               function foo(x, y)
                   return x + y
               end
           end
       end
@foo (macro with 1 method)

julia> @macroexpand @foo
quote
    #= REPL[8]:3 =#
    function var"#18#foo"(var"#19#x", var"#20#y")
        #= REPL[8]:3 =#
        #= REPL[8]:4 =#
        return var"#19#x" + var"#20#y"
    end
end

julia> @foo
#21#foo (generic function with 1 method)
```

I'm particularly skeptical that the `counter_table` can be reasonably rebuilt from the syntax of the anonymized function names alone - It seems like parsing ambiguities will be inevitable, in which case we will fail to restore the counter for one method or another.

Reply:
Wait, I thought that checking the binding table for name conflicts was precisely what you suggested [here](https://github.com/JuliaLang/julia/pull/53719#issuecomment-2047445232).

Could you clarify whether it's a misinterpretation on my end, and if so, what would be the correct interpretation of your suggestion?

Reply:
In any case, if we have a false positive in the name checking code, I believe the counter for a method might be, in the worst case, higher than it should be.

I.e., if all matches for `foo` are `#foo##0`, `#foo##1` and `#foo##42`, where the `#foo##42` is a false positive, then our counter for `foo` after re-building the table will be 43, while it should actually be 2 if we were not vulnerable to false positives.

This won't lead to correctness issues, I believe: the mangled names will have a counter that's higher than it should be, but the mangled names will still be unique.

Reply:
I don't understand why we need a duplicate table for that. If the canonical data is the bindings table, could we make an atomic "insert unique binding" here, so we don't need to do a prefix scan nor allocate a copy of a set of all prefixes, which seems still vulnerable to race conditions? (akin to having it work like mktemp)

Reply:
The false positive is in the other direction.

You might have a binding `#foo#1##1` (where `foo#1` is the name of the enclosing function). The `counter_table` will only be re-populated for `foo`, leading to a conflict when a new symbol is requested for the enclosing function `foo#1`.

Reply:
Inserting into the binding table may trigger GC (i.e. if you want to resize a `svec` which holds the bindings), which is not allowed inside `current-julia-module-counter` (marked as `JL_NOTSAFEPOINT`).

If we can relax the no-GC in `current-julia-module-counter` assumption then we may be able to possibly just keep the binding table.

My concern with using just the binding table is the potentially linear scan to check whether a canonicalized name already exists. Happy to hear if there is a way out of that.

Reply:
> You might have a binding #foo#1##1 (where foo#1 is the name of the enclosing function). The counter_table will only be re-populated for foo, leading to a conflict when a new symbol is requested for the enclosing function foo#1.

Perhaps forbid the name canonicalization if the enclosing function name contains a hash? We are already forbidding if the first character in the enclosing function name is a hash...

Reply:
Bump?

Reply:
That sounds ok to me.

Reply:
Hmm, what prevents me from defining `#foo#2` when the counter_table contains `foo => 2`, so that the next anonymized suffix for foo would be a conflict?

Reply:
You should be allowed to define something like #foo#3.

The pathological example with this implementation would be as follows:

- You have counter_table = {foo:2} and also a binding named #foo#3.
- You deserialize a module with the counter_table and binding table above.
- On deserialization, #foo#3 will show up as a false positive, in the sense that the function which rebuilds the counter_table will return {foo:3}.

Note that this example doesn’t lead to correctness issues, since the names for the deserialized module will still be unique (they will just skip one numeric suffix).

Reply:
I don't think that's a false positive, since a binding with a conflict is really there, even if it wasn't introduced via an anonymized suffix.

The case I'm worried about is when the `#foo#3` is not registered in the counter table, despite presenting a conflict in the binding table.

Specifically, what happens in #foo#3 is added after de-serialization in your scenario?